docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+
+    // Wait for up to this long for events to be available.
+    WaitTime time.Duration `json:"wait_time"`
+}
+
+type Filter struct {
+    Query string `json:"query"`
+}
+```
+
+> **Open question:** For the initial implementation, the server will not cache
+> filter queries for the client. If this turns out to be a problem, we can add
+> methods to save and remove stored queries. If necessary, the Query type can be
+> extended to include a query ID option.
+
+> **Open question:** For the initial implementation, I am planning to keep the
+> existing query language as-is. Another option would be to compile queries on
+> the client side, and provide a more structured representation in the request
+> than just a plain string.
+>
+> I think this is probably worth doing, even if only to offload the parsing
+> step onto the client. For now, though, I've stuck to the existing format to
+> make migration easier. The service can keep a small shared cache of compiled
+> queries if performance turns out to be an issue.
+
+The semantics of the request are as follows: An item in the event log is
+**eligible** for a query if:
+
+- It is newer than the `after_item` cursor (if set).
+- It is older than the `before_item` cursor (if set).
+- It matches the filter (if set).
+
+Among the eligible items in the log, the server returns up to `max_results` of
+the newest items, in reverse order of cursor. If `max_results` is unset the
+server chooses a number to return, and will cap `max_results` at a sensible
+limit.
+
+The `wait_time` parameter is used to effect polling. If `before_item` is empty,
+the server will wait for up to `wait_time` for additional items, if there are
+fewer than `max_results` eligible results in the log.  If `wait_time` is zero,
+the server will return whatever eligible items are available immediately.
+
+If `before_item` non-empty, `wait_time` is ignored: new results are only added
+to the head of the log, so there is no need to wait.  This allows the client to
+poll for new data, and "page" backward through matching event items. This is
+discussed in more detail below.
+
+The server will set a sensible cap on the maximum `wait_time`, overriding
+client-requested intervals longer than that.
+
+A successful reply from the `events` request corresponds to the following Go
+types:
+
+```go
+type EventReply struct {
+    // The items matching the request parameters, from newest
+    // to oldest, if any were available within the timeout.
+    Items []*EventItem `json:"items"`
+
+    // This is true if there is at least one older matching item
+    // available in the log that was not returned.
+    More bool `json:"more"`
+
+    // The cursor of the oldest item in the log at the time of this reply,
+    // or "" if the log is empty.
+    Oldest string `json:"oldest_item"`
+
+    // The cursor of the newest item in the log at the time of this reply,
+    // or "" if the log is empty.
+    Newest string `json:"newest_item"`
```

Comment:
are we supporting forward and reverse iteration? I sort of think that we shouldn't necessarily allow replaying within a subscription, or iterating in reverse order. 

"here's the query, here's the last ID that we got, here's the max number of items we expect to process," and have cursors be one shot and unidirectional seems like a decent (maybe have an "after" timestamp so you can say "events after now," or "don't show me events before "x" time so callers don't have to start at the beginning of cached history.

Reply:
> are we supporting forward and reverse iteration? I sort of think that we shouldn't necessarily allow replaying within a subscription, or iterating in reverse order.

Yes and no. The common case is (long-)polling at head, and if you don't get everything you page back through history. You could simulate forward paging, but it's not a first-class use case (and I don't know of a reason why we'd need it).

But there are no "subscriptions" anymore in this model, except implicit in that the client may poll/page with the same query. The server doesn't care, though.

> "here's the query, here's the last ID that we got, here's the max number of items we expect to process," and have cursors be one shot and unidirectional seems like a decent (maybe have an "after" timestamp so you can say "events after now," or "don't show me events before "x" time so callers don't have to start at the beginning of cached history.

I think I understand what you're describing here, but I'm not sure I see the motivation for making cursors behave that way, vs. having them be indices into the sliding window.

Callers never start at the beginning of cached history—the API always gives them the most recent stuff. The API is not a general-purpose timeline search—it's biased toward the specific needs of subscribers who are usually "up-to-date" and occasionally need to catch up on recent history.